nservicebus/architecture/publishing-from-web-applications.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Publishing Events from Web Applications
+summary: "Guidance on why you should or should not publish events from a web application"
+tags: []
+redirects:
+---
+
+You may have heard that [it's bad practice to publish events from a web application](http://www.make-awesome.com/2010/10/why-not-publish-nservicebus-messages-from-a-web-application/). This guidance is a bit outdated, as improvements Microsoft has made to IIS, more varied methods of hosting .NET web applications, and the addition of additional NServiceBus transports have made things much more nuanced.
+
+Yes, it is possible, and in some circumstances, even advisable to publish events from within a web application. This article will describe the guidelines for publishing messages within web applications under different circumstances.
+
+## Why/Why Not?
+
+Historically, there have been good reasons to avoid publishing events from within web applications:
+
+1. **Transactions and Consistency** - Generally good events should announce that something has *already* happened. Many times this involves making some changes to a database and then publishing the result. But HTTP is inherently unreliable and does not have built-in retries. If an exception occurs before the event is published, you may lose your only opportunity to publish that event. In these circumstances, it's generally better to send a command with the payload of the HTTP request, and have another endpoint process that command with the advantages of automatic retry and an error queue.
+2. **Web Application Scale-Out** - Because events can only be published from one logical publisher, it can be problematic to scale out a web application that is publishing events. More on this later.

```

Comment:
Perhaps out of scope for this one but we need to explain why there can be only one logical publisher and link to that page. (http://docs.particular.net/nservicebus/messaging/publish-subscribe/ doesn't seem to cover it)